File: benchmarks/lua/call/closure.lua
-- Closure benchmark
-- Tests closure creation, upvalue capture/mutation, and function composition
-- Classic funarg patterns from 1970s onward, common in production callback code

local N = tonumber(arg and arg[1]) or 10000000
collectgarbage("collect")

local function make_adder(k)
    return function(x) return x + k end
end

local function make_counter(init)
    local n = init
    return function(delta)
        n = n + delta
        return n
    end
end

local function compose(f, g)
    return function(x) return f(g(x)) end
end

local add1 = make_adder(1)
local add2 = make_adder(2)
local add3 = compose(add1, add2)

local counters = {}
for i = 1, 10 do
    counters[i] = make_counter(0)
end

local sum = 0
for i = 1, N do
    sum = sum + add3(i)

    local c = counters[(i % 10) + 1]
    sum = sum + c(1)

    if i % 10000 == 0 then
        local adder = make_adder(i)
        sum = sum + adder(i)
    end
end

print(sum)

File: benchmarks/js/call/closure.js
// Closure benchmark
// Tests closure creation, upvalue capture/mutation, and function composition
// Classic funarg patterns from 1970s onward, common in production callback code

var print = typeof console !== 'undefined' ? console.log.bind(console) : globalThis.print;
var DEFAULT = 10000000;
var N = parseInt(typeof scriptArgs !== 'undefined' ? scriptArgs[1] : (typeof process !== 'undefined' ? process.argv[2] : '0')) || DEFAULT;

function make_adder(k) {
    return function(x) { return x + k; };
}

function make_counter(init) {
    var n = init;
    return function(delta) {
        n += delta;
        return n;
    };
}

function compose(f, g) {
    return function(x) { return f(g(x)); };
}

var add1 = make_adder(1);
var add2 = make_adder(2);
var add3 = compose(add1, add2);

var counters = [];
for (var i = 0; i < 10; i++) {
    counters.push(make_counter(0));
}

var sum = 0;
for (var i = 1; i <= N; i++) {
    sum += add3(i);

    var c = counters[i % 10];
    sum += c(1);

    if (i % 10000 === 0) {
        var adder = make_adder(i);
        sum += adder(i);
    }
}

print(sum);
